src/system.rs

```diff
@@ -13,6 +14,174 @@ use std::convert::TryInto;
+        Self {
+            parent: None,
+            ids_by_name: HashMap::new(),
+            types_by_id: Vec::new(),
+        }
+    }
+
+    /// Adds all the types from Context into given TypeStore and clears Context types for loading next set of types
+    /// this is used after a schema named type/root type is loaded entirely into Context
+    pub fn update_type_store(&mut self, type_store: &SharedTypeStore) -> IonSchemaResult<()> {
+        for optional_type in &self.types_by_id {
+            // return an error if any of the type in types_by_id vector is None/Unresolved
+            let type_def = optional_type
+                .to_owned()
+                .ok_or(unresolvable_schema_error_raw(
+                    "Unable to load schema due to unresolvable type",
+                ))?;
+            match type_def.to_owned().name() {
+                Some(name) => type_store.borrow_mut().add_named_type(name, type_def),
+                None => type_store.borrow_mut().add_anonymous_type(type_def),
+            };
+        }
+        self.types_by_id.clear();
+        self.ids_by_name.clear();
+        Ok(())
+    }
+
+    /// Returns total number of types stored in the [TypeStore]
+    pub fn get_total_types(&self) -> usize {
+        self.types_by_id.len()
+    }
+
+    /// Provides the [TypeDefinition] associated with given name if it exists in the [TypeStore]  
+    /// Otherwise returns None
+    pub fn get_type_by_name(
+        &self,
+        name: &str,
+        type_store: &SharedTypeStore,
+    ) -> Option<TypeDefinition> {
+        self.ids_by_name
+            .get(name)
+            .and_then(|id| match self.types_by_id.get(*id) {
+                Some(t) => Some(t.to_owned().unwrap()),
+                None => match type_store.borrow().get_type_by_id(id.to_owned()) {
+                    Some(t) => Some(t.to_owned()),
+                    None => None,
+                },
+            })
+    }
+
+    /// Provides the [TypeId] associated with given name if it exists in the [TypeStore] or [Context]  
+    /// Otherwise returns None
+    pub fn get_type_id_by_name(&self, name: &str, type_store: &SharedTypeStore) -> Option<TypeId> {
+        match self.ids_by_name.get(name) {
+            Some(id) => Some(id.to_owned()),
+            None => match type_store.borrow().get_type_id_by_name(name) {
+                Some(id) => Some(id.to_owned()),
+                None => None,
+            },
+        }
+    }
+
+    /// Provides the [TypeDefinition] associated with given [TypeId] if it exists in the [TypeStore]  
+    /// or in the [Context] Otherwise returns None
+    pub fn get_type_by_id(
+        &self,
+        id: TypeId,
+        type_store: &SharedTypeStore,
+    ) -> Option<TypeDefinition> {
+        match self.types_by_id.get(id) {
+            Some(t) => Some(t.to_owned().unwrap()),
+            None => match type_store.borrow().get_type_by_id(id) {
+                Some(t) => Some(t.to_owned()),
```

Comment:
I see a lot of `TypeDefinition` cloning happening in this file that I think we can eliminate in another PR.

I think making `Context` responsible for the fallback-to-`TypeStore` logic is muddying the waters a bit. IMO, `get_type_by_id` should only return the type if the `Context` has it. Callers can instead do:

```rust
  self.context
    .get_type_by_id(type_id)
    .or_else(|| type_store.borrow().get_type_by_id(type_id))
```

Reply:
The issue here is that `Context` stores `TypeDefintion` in terms of `Option` i.e. `types_by_id: Vec<Option<TypeDefinition>>`, where as TypeStore stores TypeDefinition directly i.e. `types_by_id: Vec<TypeDefinition>`. Because of this I would have to return `Option<TypeDefintion>` here rather than `Option<&TypeDefinition>`.  I could make it concise by using `or_else` and `and_then` but `to_owned()` still needs to be used. Also, if I try to do something like below:
```rust
 self.context
    .get_type_by_id(type_id).and_then(|t| &t.unwrap())   // <---- this &__t.unwrap()__ creates a temporary value that can not be returned from this fucntion.
    .or_else(|| type_store.borrow().get_type_by_id(type_id)
```

Reply:
Why does the `Context` ever need to _give out_ a `TypeDefinition`? I thought it just accumulated `TypeDefinition`s and then moved them all into the `TypeStore` at the end of the resolution process.

Reply:
I am using the `Context` while resolving a type inside `resolve_type_reference` where I refer to the `types_by_id` vector inside the `Context` to check if a resolved `TypeDefinition` already exists.

Reply:
Upon having another look at the method and from offline discussion the method isn't really required to be here inside `Context` struct hence will be removing this method.